### ISSUE TITLE
Could not install tags that are not listed in the package.py anymore

### ISSUE THREAD

**Issue by uturuncoglu** [LLNL/spack#37889]
### Steps to reproduce

Spack is not allowing me to install tags that are not listed in the package.py anymore after PR https://github.com/spack/spack/pull/37438. I am not sure this is expected behavior or not but it would be nice to bring this capability again to spack or fix it. 

To reproduce the issue, tying to installing a version of package that is not listed in the package.py is enough. At this point I tried with two different packages like `esmf` and `netcdf-c` and both gives same issue

```
$ spack install --no-checksum  esmf@8.5.0b10
```

or

```
spack install --no-checksum netcdf-c@4.9.1
```

and it will give following error,

```
==> Error: concretization failed for the following reasons:

   1. Cannot satisfy 'esmf@v8.5.0b10'
```

In both cases, either `8.5.0b10` for `esmf` or `4.9.1` for `netcdf-c` are not listed as version in the package.py. These commands are working if I checkout the version of Spack just before the PR (git checkout b2c3973).

### Error message

```
$ spack --debug --stacktrace install --no-checksum netcdf-c@4.9.1
lib/spack/spack/cmd/__init__.py:123 ==> [2023-05-24-17:00:46.056679] Imported install from built-in commands
lib/spack/spack/config.py:996 ==> [2023-05-24-17:00:46.066607] Reading config from file /test/spack/etc/spack/defaults/config.yaml
lib/spack/spack/config.py:996 ==> [2023-05-24-17:00:46.104832] Reading config from file /root/.spack/config.yaml
lib/spack/spack/cmd/__init__.py:123 ==> [2023-05-24-17:00:46.107848] Imported install from built-in commands
lib/spack/spack/config.py:996 ==> [2023-05-24-17:00:46.124319] Reading config from file /test/spack/etc/spack/defaults/bootstrap.yaml
lib/spack/spack/config.py:996 ==> [2023-05-24-17:00:46.139551] Reading config from file /test/spack/etc/spack/defaults/repos.yaml
lib/spack/spack/config.py:996 ==> [2023-05-24-17:00:46.146685] Reading config from file /test/spack/etc/spack/defaults/config.yaml
lib/spack/spack/config.py:996 ==> [2023-05-24-17:00:46.184948] Reading config from file /root/.spack/config.yaml
lib/spack/spack/database.py:391 ==> [2023-05-24-17:00:46.188302] DATABASE LOCK TIMEOUT: 60s
lib/spack/spack/database.py:397 ==> [2023-05-24-17:00:46.188764] PACKAGE LOCK TIMEOUT: No timeout
lib/spack/spack/database.py:391 ==> [2023-05-24-17:00:46.189997] DATABASE LOCK TIMEOUT: 60s
lib/spack/spack/database.py:397 ==> [2023-05-24-17:00:46.190360] PACKAGE LOCK TIMEOUT: No timeout
lib/spack/spack/bootstrap/config.py:125 ==> [2023-05-24-17:00:46.190770] [BOOTSTRAP CONFIG SCOPE] name=_builtin
lib/spack/spack/config.py:996 ==> [2023-05-24-17:00:46.192046] Reading config from file /test/spack/etc/spack/defaults/bootstrap.yaml
lib/spack/spack/bootstrap/config.py:138 ==> [2023-05-24-17:00:46.203700] [BOOTSTRAP CONFIG SCOPE] name=defaults, path=/test/spack/etc/spack/defaults
lib/spack/spack/bootstrap/config.py:139 ==> [2023-05-24-17:00:46.204119] [BOOTSTRAP CONFIG SCOPE] name=defaults/linux, path=/test/spack/etc/spack/defaults/linux
lib/spack/spack/bootstrap/config.py:138 ==> [2023-05-24-17:00:46.204390] [BOOTSTRAP CONFIG SCOPE] name=bootstrap, path=/root/.spack/bootstrap/config
lib/spack/spack/bootstrap/config.py:139 ==> [2023-05-24-17:00:46.204758] [BOOTSTRAP CONFIG SCOPE] name=bootstrap/linux, path=/root/.spack/bootstrap/config/linux
lib/spack/spack/config.py:996 ==> [2023-05-24-17:00:46.206306] Reading config from file /root/.spack/bootstrap/config/linux/compilers.yaml
lib/spack/spack/bootstrap/_common.py:223 ==> [2023-05-24-17:00:46.267530] [BOOTSTRAP ROOT SPEC] patchelf@0.13.1: %gcc target=x86_64
lib/spack/spack/config.py:996 ==> [2023-05-24-17:00:46.269045] Reading config from file /test/spack/etc/spack/defaults/bootstrap.yaml
lib/spack/spack/bootstrap/_common.py:185 ==> [2023-05-24-17:00:46.298270] [BOOTSTRAP EXECUTABLES patchelf] Try installed specs with query 'patchelf@0.13.1: %gcc target=x86_64'
lib/spack/spack/config.py:996 ==> [2023-05-24-17:00:46.391256] Reading config from file /test/spack/etc/spack/defaults/config.yaml
lib/spack/spack/config.py:996 ==> [2023-05-24-17:00:46.495625] Reading config from file /test/spack/etc/spack/defaults/modules.yaml
lib/spack/spack/config.py:996 ==> [2023-05-24-17:00:46.514806] Reading config from file /test/spack/etc/spack/defaults/linux/modules.yaml
lib/spack/spack/config.py:996 ==> [2023-05-24-17:00:46.617294] Reading config from file /test/spack/etc/spack/defaults/packages.yaml
lib/spack/spack/bootstrap/_common.py:223 ==> [2023-05-24-17:00:46.676359] [BOOTSTRAP ROOT SPEC] gnupg@2.3: %gcc target=x86_64
lib/spack/spack/bootstrap/_common.py:223 ==> [2023-05-24-17:00:46.681020] [BOOTSTRAP ROOT SPEC] clingo-bootstrap@spack+python %gcc target=x86_64
lib/spack/spack/bootstrap/core.py:376 ==> [2023-05-24-17:00:46.681498] [BOOTSTRAP MODULE clingo] Try importing from Python
lib/spack/spack/bootstrap/_common.py:78 ==> [2023-05-24-17:00:46.713116] [BOOTSTRAP MODULE clingo] The installed spec "clingo-bootstrap@spack+python %gcc target=x86_64 ^python@3.10/idkenmhnscjlu5gjqhpcqa4h7o2a7aow" provides the "clingo" Python module
lib/spack/spack/config.py:996 ==> [2023-05-24-17:00:46.714394] Reading config from file /test/spack/etc/spack/defaults/concretizer.yaml
lib/spack/spack/config.py:996 ==> [2023-05-24-17:00:46.729512] Reading config from file /test/spack/etc/spack/defaults/repos.yaml
lib/spack/spack/config.py:996 ==> [2023-05-24-17:00:46.861916] Reading config from file /test/spack/etc/spack/defaults/packages.yaml
lib/spack/spack/config.py:996 ==> [2023-05-24-17:00:46.925297] Reading config from file /root/.spack/packages.yaml
lib/spack/spack/config.py:996 ==> [2023-05-24-17:00:47.008988] Reading config from file /test/spack/etc/spack/defaults/config.yaml
lib/spack/spack/config.py:996 ==> [2023-05-24-17:00:47.060388] Reading config from file /root/.spack/config.yaml
lib/spack/spack/config.py:996 ==> [2023-05-24-17:00:47.071166] Reading config from file /test/spack/etc/spack/defaults/mirrors.yaml
lib/spack/spack/util/web.py:345 ==> [2023-05-24-17:00:47.081055] Checking existence of https://mirror.spack.io/build_cache/index.json
lib/spack/spack/util/executable.py:195 ==> [2023-05-24-17:00:47.086194] '/usr/bin/curl' '--stderr' '-' '-s' '-f' '-r' '0-0' 'https://mirror.spack.io/build_cache/index.json'
lib/spack/spack/config.py:996 ==> [2023-05-24-17:00:50.076629] Reading config from file /root/.spack/linux/compilers.yaml
lib/spack/spack/cmd/install.py:452 ==> [2023-05-24-17:01:18.371395] SpackError: concretization failed for the following reasons:

   1. Cannot satisfy 'netcdf-c@4.9.1'
lib/spack/spack/main.py:1021 ==> [2023-05-24-17:01:18.371721] SpackError: concretization failed for the following reasons:

   1. Cannot satisfy 'netcdf-c@4.9.1'
lib/spack/spack/error.py:54 ==> [2023-05-24-17:01:18.372618] Error: concretization failed for the following reasons:

   1. Cannot satisfy 'netcdf-c@4.9.1'
Traceback (most recent call last):
  File "/test/spack/lib/spack/spack/cmd/__init__.py", line 220, in parse_specs
    spec.concretize(tests=tests)  # implies normalize
  File "/test/spack/lib/spack/spack/spec.py", line 2915, in concretize
    self._new_concretize(tests)
  File "/test/spack/lib/spack/spack/spec.py", line 2890, in _new_concretize
    result = solver.solve([self], tests=tests)
  File "/test/spack/lib/spack/spack/solver/asp.py", line 2774, in solve
    result, _, _ = self.driver.solve(setup, specs, reuse=reusable_specs, output=output)
  File "/test/spack/lib/spack/spack/solver/asp.py", line 817, in solve
    error_handler.raise_if_errors()
  File "/test/spack/lib/spack/spack/solver/asp.py", line 651, in raise_if_errors
    raise UnsatisfiableSpecError(msg)
spack.solver.asp.UnsatisfiableSpecError: concretization failed for the following reasons:

   1. Cannot satisfy 'netcdf-c@4.9.1'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/test/spack/lib/spack/spack/main.py", line 1018, in main
    return _main(argv)
  File "/test/spack/lib/spack/spack/main.py", line 973, in _main
    return finish_parse_and_run(parser, cmd_name, env_format_error)
  File "/test/spack/lib/spack/spack/main.py", line 1001, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
  File "/test/spack/lib/spack/spack/main.py", line 650, in _invoke_command
    return_val = command(parser, args)
  File "/test/spack/lib/spack/spack/cmd/install.py", line 368, in install
    install_without_active_env(args, install_kwargs, reporter_factory)
  File "/test/spack/lib/spack/spack/cmd/install.py", line 481, in install_without_active_env
    concrete_specs = concrete_specs_from_cli(args, install_kwargs) + concrete_specs_from_file(args)
  File "/test/spack/lib/spack/spack/cmd/install.py", line 448, in concrete_specs_from_cli
    concrete_specs = spack.cmd.parse_specs(
  File "/test/spack/lib/spack/spack/cmd/__init__.py", line 234, in parse_specs
    raise spack.error.SpackError(msg) from e
spack.error.SpackError: concretization failed for the following reasons:

   1. Cannot satisfy 'netcdf-c@4.9.1'
```

### Information on your system

* **Spack:** 0.21.0.dev0 (785c31b730586045f5d9ea13fc61ed72b8da7d8e)
* **Python:** 3.10.6
* **Platform:** linux-ubuntu22.04-haswell
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by uturuncoglu** [comment]
@becker33 @alalazo @tgamblin I know you are probably busy with other things but do you have any idea about this issue?

**Comment by alalazo** [comment]
**TL;DR** If you use `esmf@=8.5.0b10` the spec should concretize

----

This is a consequence of #36273 In short:
- `@X.Y` always denotes the version range `@X.Y:X.Y`
- `@=X.Y` is instead the specific `X.Y` version

Before the PR above the two meanings were ambiguous[^1]. A consequence of the new, more precise, semantic is that for versions that disappeared you have to use the `@=` syntax to let Spack know you really mean that specific version. 

Otherwise, Spack will interpret the token as a version range, and might end up in a spot where that version range is an empty set.

[^1]: think of `mpich`, which has versions with a different number of digits `3.0`, `3.0.1`, ... Before there was no good way to specify _exactly 3.0_

**Comment by alalazo** [comment]
Marked as not a bug, but let me know if there are further issues.

**Comment by uturuncoglu** [comment]
@alalazo Thanks. It seems that new syntax is working for me. Thanks for your help.
